[Update 8] - Add breaking changes for U8 (#7642)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.8.0/RELEASE_NOTE.md

## Backward-incompatible changes

> - Fixed a bug that resulted in variables that may or may not be initialized in an `on fail` clause not being identified as potentially uninitialized variables.

Reviewer: I believe we decided to use the format "A bug that resulted in ... has been fixed" (e.g., https://ballerina.io/downloads/swan-lake-release-notes/swan-lake-2201.7.0).

To avoid reading as the bug fix is resulting in the particular behaviour.

  ```ballerina

>   public function main() {

Reviewer 2: suggestion: better if we can add the import in here
`import ballerina/io;`

    int resultInt;
    transaction {
        resultInt = check maybeProvideInt(true);
        check commit;
    } on fail {
        io:println("Failed to initialize resultInt");
    }
    resultInt += 1; // error: resultInt may not have been initialized
  }
  ```

> - Fixed a bug related to type inference within query expressions when an expected type was absent. Previously, when iterating over a `map` without specifying an expected type, the query expression's result type was mistakenly inferred as an `array` which is now restricted.

Reviewer: Here map and array refer to map and array values, right? Not the keyword (at least for map).

We only emphasize keywords, variable names, etc. and that too when referred to as keywords, names, etc. I don't think we need to emphasize map here. We shouldn't emphasize array.

  
- ```ballerina
  function foo(map<int> mp) {

>       var _ = from int i in mp // compile-time error

Reviewer: For examples, we decided not to use `_`, right?

>       var _ = from int i in mp // compile-time error

Reviewer: Shall we make this a meaningful example? Without `foo` and using a real-world name for the map.

                select i;
  }
  ```

> - Fixed a bug where `error` completion was ignored while using collect-clause with query expressions.

Reviewer: ```suggestion
- Fixed a bug where possible completion with `error` was ignored when using `collect` clause in a query expression.
```

 
- ```ballerina

>   function foo(stream<int, error?> strm) {

Reviewer: Same comments as above.

- Ballerina interoperability implementation may have an impact with the Java 17 support due to any incompatible changes. For example, Java 17 has some restrictions on using Java reflections with internal Java packages. For more information, see the Java 17 release notes.

> - A bug that allowed variables within an `on fail` clause, which may or may not have been initialized, to evade detection as potentially uninitialized variables, has been fixed.

Reviewer: Is this correct? Not exactly "within" an on fail clause?

Author: Rephrased. Please check

  ```ballerina
  public function main() {
    int resultInt;
    transaction {
        resultInt = check maybeProvideInt(true);
        check commit;
    } on fail {
        io:println("Failed to initialize resultInt");
    }
    resultInt += 1; // error: resultInt may not have been initialized
  }
  ```

- A bug that allowed incorrect type inference within query expressions when an expected type was absent has been addressed. Previously, when iterating over a map without explicitly specifying an expected type, the resulting type of the query expression was erroneously inferred as an array. This misinterpretation has now been rectified and is properly restricted.
  
- ```ballerina
  function iterateMap(map<int> mp) {

>       var result = from int i in mp // compile-time error

Reviewer: I think we use "Compilation error now.".

Author: Addressed in d4d8baf1d0d2cebb74a6c9d69ba0a4080805ad45

>                 select i;

Reviewer: I would make this more real-world, like filtering with a simple where clause (and change the function name accordingly).

Author: Addressed in d4d8baf1d0d2cebb74a6c9d69ba0a4080805ad45

  }
  ```

- A bug that allowed ignoring possible completion with an error when using the `collect` clause in a query expression has been fixed.
 
- ```ballerina
  function calculateTotal(stream<int, error?> strm) {

>      int _ = from var i in strm // compile-time error: expected int, but found int|error

Reviewer: Still uses `_`.

Author: Addressed in d4d8baf1d0d2cebb74a6c9d69ba0a4080805ad45

- Ballerina interoperability implementation may have an impact with the Java 17 support due to any incompatible changes. For example, Java 17 has some restrictions on using Java reflections with internal Java packages. For more information, see the Java 17 release notes.

- A bug that permitted uninitialized variables to evade detection when utilizing the `on fail` clause has been fixed.

  ```ballerina
  public function main() {
    int resultInt;
    transaction {

>         resultInt = check maybeProvideInt(true);

Reviewer: `getOptionalInt` or something more real-world? Even something like `getCutOff` or min max?

Author: Addressed

        check commit;
    } on fail {
        io:println("Failed to initialize resultInt");
    }
    resultInt += 1; // Compilation error now.
  }
  ```

> - A bug that allowed incorrect type inference within query expressions when an expected type was absent has been addressed. Previously, when iterating over a map without explicitly specifying an expected type, the resulting type of the query expression was erroneously inferred as an array. This misinterpretation has now been rectified and is properly restricted.

Reviewer: caused or resulted in instead of allowed?

when there is no expected type instead of when an expected type was absent?

Author: Addressed
